docs/userguide/src/portingguide/topics/weakref.md

```diff
@@ -0,0 +1,451 @@
+# Finalizers and Weak References
+
+Some VMs support **finalizers**.  In simple terms, finalizers are clean-up operations associated
+with an object, and are executed when the object is dead.
+
+Some VMs support **weak references**.  If an object cannot be reached from roots following only
+strong references, the object will be considered dead.  Weak references to dead objects will be
+cleared, and associated clean-up operations will be executed.  Some VMs also support more complex
+weak data structures, such as weak hash tables, where keys, values, or both, can be weak references.
+
+The concrete semantics of finalizer and weak reference varies from VM to VM, but MMTk provides a
+low-level API that allows the VM bindings to implement their flavors of finalizer and weak
+references on top of it.
+
+**A note for Java programmers**: In Java, the term "weak reference" often refers to instances of
+`java.lang.ref.Reference` (including the concrete classes `SoftReference`, `WeakReference`,
+`PhantomReference` and the hidden `FinalizerReference` class used by some JVM implementations to
+implement finalizers).  Instances of `Reference` are proper Java heap objects, but each instance has
+a field that contains a pointer to the referent, and the field can be cleared when the referent
+dies.  In this article, we use the term "weak reference" to refer to the pointer inside that field.
+In other words, a Java `Reference` instance has a field that holds a weak reference to the referent.
+
+## Overview
+
+During each GC, after the transitive closure is computed (i.e. after all objects reachable from
+roots have been reached), MMTk calls `Scanning::process_weak_refs` which is implemented by the VM
+binding.  Inside this function, the VM binding can do several things.
+
+-   **Query reachability**: The VM binding can query whether any given object has been reached.
+    +   Do this with `ObjectReference::is_reachable()`.
+-   **Query forwarded address**: If an object is already reached, the VM binding can further query
+    the new address of an object.  This is needed to support copying GC.
+    +   Do this with `ObjectReference::get_forwarded_object()`.
+-   **Resurrect objects**: If an object is not reached, the VM binding can optionally resurrect the
```

Comment:
See my argument above for the definition of weak ref. We do not 'resurect' weak reference. 

It is arguable whether we should say 'resurrect' finalizable objects. GC needs to keep finalizable objects alive until they are properly finalized -- I don't want to call it 'resurrect', simply it is not dead from the GC's perspective. Also 'resurrecting objects' normally refers to the specific application behavior that an object that is being finalized and should become dead after finalization is at the language/application level kept alive. Using 'resurrect' in the GC may get things confusing.

GC handbook's definition for 'resurrection': "an action performed by a finaliser that caused the previously unreachable object to become reachable."

Reply:
I consulted @eliotmoss and read the GC handbook, and they both agree with what you said.  "Resurrection" refers to the action in the finalizer that makes the object reachable from other parts of the program by, for example, assigning its reference to a global variable.  This is not what `tracer.trace_object` does.  Even if we call `trace_object(object)`, it still doesn't guarantee the `object` will "resurrect" in that sense.  If the `finalize()` function doesn't leak its reference, the object will still be collectable after `finalize()` returns.

"Retain" should be a better term, and it is already used in JikesRVM as well as our existing `ReferenceProcessor` in mmtk-core.

Reply:
I now use "retain" consistently when describing what `tracer.trace_object` does.  I still mentioned "resurrect" briefly when describing the counterintuitive behavior that finalizers make an unreachable object reachable from user threads again.